src/nvim/os/env.c

```diff
@@ -169,6 +178,64 @@ void os_get_hostname(char *hostname, size_t size)
 #endif
 }
 
+/// @return -1 if error encountered, 1 if capslock active otherwise 0 if not.
+int capslock_status(void ) {
+
+#if defined(__linux__)
+
+  FILE *fp = NULL;
+  char res[1024];
+  fp = popen("xset -q | grep Caps", "r");
+  if (fp == NULL) {
+      return -1;
+  }
+  fgets(res, sizeof(res), fp);
+  if (STRNCMP(res + 21, "on", 2) == 0) {
+    return 1;
+  } else {
+    return 0;
+  }
+
+#elif defined(WIN32)
+  if ((GetKeyState(VK_CAPITAL) & 0x0001)!=0) {
+    return 1
+  } else {
+    return 0;
+  }
+#elif defined(__APPLE__)
+
+  kern_return_t kr;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool state;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault, (CFDictionaryRef) mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    return -1;
+  }
+
+  kr = IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType, &ioc);
```

Comment:
kr is not a pointer but a variable, simple variable, so its space will be allocated on stack, and I think due to the comparsons made, it may be like some integer, but not sure, I will check it

Reply:
File descriptors also fit into the definition of “simple variable, not a pointer”, yet they do need to be closed. Would not expect that for a variable with type named `kern_return_t` though, it looks more like errno, especially after seeing it checked agaist something like `…_SUCCESS`. Though unless you are going to do with `kr` something other than checking against `KERN_SUCCESS` once (that is, if `kr` cannot be used to create useful error messages or you are not going to do so) I would suggest to remove the variable completely and put calls into the `if`.

Reply:
Though no, based on the code you can only remove the second `kr`. Still that makes first perfect to have `const` in the type.